File: src/main/java/cz/astro/var/data/security/CustomUserDetailsService.java
package cz.astro.var.data.security;

import cz.astro.var.data.czev.repository.User;
import cz.astro.var.data.czev.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.stream.Collectors;

@Service
public class CustomUserDetailsService implements UserDetailsService {

    @Autowired
    private UserRepository userRepository;

    @Override
    @Transactional
    public UserPrincipal loadUserByUsername(String username) throws UsernameNotFoundException {
        User user = userRepository.findByEmailFetched(username).orElseThrow(() -> new UsernameNotFoundException("User not found"));

        UserPrincipal userPrincipal = new UserPrincipal();
        userPrincipal.setEmail(user.getEmail());
        userPrincipal.setPassword(user.getPassword());
        userPrincipal.setId(user.getId());
        userPrincipal.setAuthorities(user.getRoles().stream()
                .map(r -> new SimpleGrantedAuthority(r.getName())).collect(Collectors.toList()));

        return userPrincipal;
    }

    @Transactional
    public UserPrincipal loadUserById(Long id) {
        User user = userRepository.findByIdFetched(id).orElseThrow(() -> new UsernameNotFoundException("User not found"));

        UserPrincipal userPrincipal = new UserPrincipal();
        userPrincipal.setEmail(user.getEmail());
        userPrincipal.setPassword(user.getPassword());
        userPrincipal.setId(user.getId());
        userPrincipal.setAuthorities(user.getRoles().stream()
                .map(r -> new SimpleGrantedAuthority(r.getName())).collect(Collectors.toList()));

        return userPrincipal;
    }
}

File: backend/src/main/kotlin/cz/astro/var/data/security/CustomUserDetailsService.kt
package cz.astro.`var`.data.security

import cz.astro.`var`.data.czev.repository.UserRepository
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class CustomUserDetailsService(
        private val userRepository: UserRepository
) : UserDetailsService {

    @Transactional
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserPrincipal {
        val user = userRepository.findByEmailFetched(username).orElseThrow { UsernameNotFoundException("User not found") }

        val userPrincipal = UserPrincipal()
        userPrincipal.email = user.email
        userPrincipal.setPassword(user.password)
        userPrincipal.id = user.id
        userPrincipal.setAuthorities(user.roles
                .map { r -> SimpleGrantedAuthority(r.name) }.toList())

        return userPrincipal
    }

    @Transactional
    fun loadUserById(id: Long?): UserPrincipal {
        val user = userRepository.findByIdFetched(id!!).orElseThrow { UsernameNotFoundException("User not found") }

        val userPrincipal = UserPrincipal()
        userPrincipal.email = user.email
        userPrincipal.setPassword(user.password)
        userPrincipal.id = user.id
        userPrincipal.setAuthorities(user.roles
                .map { r -> SimpleGrantedAuthority(r.name) }.toList())

        return userPrincipal
    }
}
